Detect invalid ID to H5Gmove2

bmribler opened the pull request:
User's application segfaulted because the returned value H5I_BADID wasn't detected when H5I_get_type() was called.  This PR adds checks for invalid file/group identifiers passed into H5Gmove2.

Question: Is that still necessary to check the returned value from H5I_get_type() or should the fix be checking the returned value from H5I_get_type() instead?

This defect occurs in many other places, hence, issue GH-4764.

Fixes GH-4737

bmribler committed:
Detect invalid ID to H5Gmove2

User's application segfaulted because the returned value H5I_BADID wasn't
detected when H5I_get_type() was called.  This PR adds checks for invalid
file/group identifiers passed into H5Gmove2.

Is that still necessary to check the returned value from H5I_get_type() or
should the fix be checking the returned value from H5I_get_type() instead?

This defect occurs in many other places, hence, issue GH-4764.

Fixes GH-4737

github-actions[bot] committed:
Committing clang-format changes

derobins commented on test/links.c:
```diff
@@ -2117,6 +2117,22 @@ test_deprec(hid_t fapl, bool new_format)
+    /* Test passing in invalid ID */
+    H5E_BEGIN_TRY
+    {
+        if (H5Gmove2(0, "group2", group1_id, "moved_group2") >= 0)
```
I would make an explicit bad_id = H5I_INVALID_HID variable and pass that in, instead of using a normal ID that happens to be set to H5I_INVALID_HID. If someone reworks the test, that might get set.

bmribler replied:
Actually, this test is meant to detect 0 being passed in, like in the user's sample program.  I can add what you said though.

bmribler replied:
Nevermind, 0 is the case of H5L_SAME_LOC.  I'll have to redo this.

qkoziol commented on src/H5Gdeprec.c:
```diff
@@ -519,10 +519,19 @@ H5Gmove2(hid_t src_loc_id, const char *src_name, hid_t dst_loc_id, const char *d
 
     FUNC_ENTER_API(FAIL)
 
+    id_type = H5I_get_type(src_loc_id);
```
You will need to handle allowing the magic H5L_SAME_LOC values for the src & dst IDs.

bmribler replied:
Oh ok, hmm, so if the caller passes in 0, we assume they meant H5L_SAME_LOC?

bmribler replied:
Also, the function has already handled allowing the magic H5L_SAME_LOC values, except that
loc_params1.obj_type                     = H5I_get_type(src_loc_id);  <- later causing segfault

happens before

    if (H5L_SAME_LOC != src_loc_id)
        /* get the location object */

bmribler committed:
Replace non-VOL calls with VOL calls - part 5

This PR switches H5I_object() and H5I_object_verify() to H5VL_vol_object() and          H5VL_vol_object_verify(), respectively, in the H5M APIs and H5Gdeprec (was left         out by mistake).  This completes the fixes of issue GH-4730.

bmribler committed:
Merge remote-tracking branch 'origin/fix_GH-4730_part5' into fix_GH-4737

bmribler committed:
Fix how H5L_SAME_LOC is handled.
                                                                                   Reset an ID to the other ID when it is H5L_SAME_LOC, then we will not
have to deal with NULL vol_obj later.

github-actions[bot] committed:
Committing clang-format changes

bmribler committed:
Merge branch 'develop' into fix_GH-4737

bmribler commented:
I made requested changes.  Thanks!

qkoziol reviewed (approved):
LGTM :-)
